File: leecode/leetcode-cn/25.k个一组翻转链表.java

/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
  public ListNode reverseKGroup(ListNode head, int k) {
      ListNode dummy = new ListNode();
      dummy.next = head;

      ListNode prev = dummy;

      while (prev != null) {
          ListNode left = prev.next;

          // 每次走k-1步，到达right
          ListNode right = left;
          for (int i = 0; i < k - 1 && right != null; i++)
              right = right.next;

          if (right == null)
              break;

          ListNode next = right.next;

          // 截断
          prev.next = null;
          right.next = null;

          reverseList(left);

          // 重新拼接
          prev.next = right;
          left.next = next;

          // 下一次循环
          prev = left;
      }

      return dummy.next;
  }

  public void reverseList(ListNode head) {
      if (head == null || head.next == null)
          return;

      reverseList(head.next);
      head.next.next = head;
      head.next = null;
  }
}

File: leecode/leetcode-cn/25.k个一组翻转链表.cc
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode* dummy = new ListNode();
        dummy->next = head;
        ListNode* prev = dummy;
        while (prev) {
            // 将 [left, right] 范围内的链表翻转
            ListNode* left = prev->next, *right = left;
            // 移动 k - 1 次，则 right 恰好指向第 k 个节点
            for (int i = 0; i < k - 1 && right != nullptr; ++i)
                right = right->next;
            // 说明节点不足 k 个，不需要翻转了
            if (right == nullptr)
                break;
            // 暂存 next
            ListNode* next = right->next;
            // 将 [left, right] 子链表截断
            prev->next = nullptr;
            right->next = nullptr;
            // 翻转子链表
            reverseList(left);
            // 重新拼接
            prev->next = right;
            left->next = next;
            // 下一次循环
            prev = left;
        }
        return dummy->next;
    }

    ListNode* reverseList(ListNode* head) {
        if (head == nullptr || head->next == nullptr)
            return head;
        ListNode* p = reverseList(head->next);
        head->next->next = head;
        head->next = nullptr;
        return p;
    }
};

// 方法二：官方题解
class Solution {
public:
    // 翻转一个子链表，并且返回新的头与尾
    pair<ListNode*, ListNode*> myReverse(ListNode* head, ListNode* tail) {
        ListNode* prev = tail->next;
        ListNode* p = head;
        while (prev != tail) {
            ListNode* nex = p->next;
            p->next = prev;
            prev = p;
            p = nex;
        }
        return {tail, head};
    }

    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode* hair = new ListNode(0);
        hair->next = head;
        ListNode* pre = hair;

        while (head) {
            ListNode* tail = pre;
            // 查看剩余部分长度是否大于等于 k
            for (int i = 0; i < k; ++i) {
                tail = tail->next;
                if (!tail) {
                    return hair->next;
                }
            }
            ListNode* nex = tail->next;
            // 这里是 C++17 的写法，也可以写成
            // pair<ListNode*, ListNode*> result = myReverse(head, tail);
            // head = result.first;
            // tail = result.second;
            tie(head, tail) = myReverse(head, tail);
            // 把子链表重新接回原链表
            pre->next = head;
            tail->next = nex;
            pre = tail;
            head = tail->next;
        }

        return hair->next;
    }
};

// 作者：LeetCode-Solution
// 链接：https://leetcode.cn/problems/reverse-nodes-in-k-group/solution/k-ge-yi-zu-fan-zhuan-lian-biao-by-leetcode-solutio/
// 来源：力扣（LeetCode）
// 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。